### PR TITLE
Add support for SstFileWriter and DB::ingest_external_file

### PR DESCRIPTION
**Pull request by methyl**

### CODE REVIEW

File: src/db.rs [hunk 1]
```diff
@@ -1107,6 +1107,86 @@ impl DB {
         }
         Ok(())
     }
+
+    pub fn ingest_external_file<P: AsRef<Path>>(&self, paths: Vec<P>) -> Result<(), Error> {
```
Comment (aleksuss) [inline src/db.rs:1111]: It would be great to add some description to the public methods.

File: src/sst_file_writer.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//`
+
+use crate::{ffi, ffi_util::to_cpath, Error, Options};
+
+use libc::{self, c_char, size_t};
+use std::ffi::CString;
+use std::marker::PhantomData;
+use std::path::Path;
```
Comment (aleksuss) [inline src/sst_file_writer.rs:21]: Make `std` imports nested please.

Comment (methyl) [reply src/sst_file_writer.rs:21]: from what i've seen you never do that in other files, can you show the desired effect?

Comment (methyl) [reply src/sst_file_writer.rs]: If [this](https://github.com/rust-rocksdb/rust-rocksdb/pull/421/commits/2ff1d0d2e6fa0ec6baeeb1e3d66e4ba3d89c8bc6) is what you meant, I don't think it makes sense to do that only for this particular file

Comment (aleksuss) [reply src/sst_file_writer.rs]: Yes. That's exactly what I mean. I'll change it in other places later.

File: src/db.rs [hunk 3]
```diff
@@ -1107,6 +1107,91 @@ impl DB {
...
+            .iter()
+            .map(|path| to_cpath(&path))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let cpaths: Vec<_> = paths_v.iter().map(|path| path.as_ptr()).collect();
+
+        self.ingest_external_file_raw_cf(&cf, &opts, &paths_v, &cpaths)
+    }
+
+    pub fn ingest_external_file_raw(
```
Comment (aleksuss) [inline src/db.rs:1161]: Does it make sense to make this method public ???

File: src/db.rs [hunk 4]
```diff
@@ -1107,6 +1107,91 @@ impl DB {
...
+            ffi_try!(ffi::rocksdb_ingest_external_file(
+                self.inner,
+                cpaths.as_ptr(),
+                paths_v.len(),
+                opts.inner as *const _
+            ));
+            Ok(())
+        }
+    }
+    pub fn ingest_external_file_raw_cf(
```
Comment (aleksuss) [inline src/db.rs:1177]: Same here. Private visibility is enough I suppose.

File: src/db_options.rs [hunk 5]
```diff
@@ -1862,6 +1874,50 @@ impl Default for ReadOptions {
     }
 }
 
+impl IngestExternalFileOptions {
+    pub fn set_move_files(&mut self, v: bool) {
```
Comment (aleksuss) [inline src/db_options.rs:1878]: A description for the public method is missing.

**Comment by methyl** [comment]
@aleksuss let's crush it :muscle:

File: src/sst_file_writer.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    pub(crate) inner: *mut ffi::rocksdb_sstfilewriter_t,
+    // Options are needed to be alive when calling open(),
+    // so let's make sure it doesn't get, dropped for the lifetime of SstFileWriter
+    phantom: PhantomData<&'a Options>,
+}
+
+unsafe impl<'a> Send for SstFileWriter<'a> {}
+unsafe impl<'a> Sync for SstFileWriter<'a> {}
+
+pub struct EnvOptions {
```
Comment (aleksuss) [inline src/sst_file_writer.rs:33]: Description ?

Comment (methyl) [reply src/sst_file_writer.rs:33]: this one should actually private, there is no way to customize it - rocksdb hardcodes the constructor call everywhere. Ill just unpublish it

**Comment by methyl** [comment]
@aleksuss I guess everything has been addressed

File: src/db.rs [hunk 7]
```diff
@@ -1107,6 +1107,91 @@ impl DB {
...
+        unsafe {
+            ffi_try!(ffi::rocksdb_ingest_external_file(
+                self.inner,
+                cpaths.as_ptr(),
+                paths_v.len(),
+                opts.inner as *const _
+            ));
+            Ok(())
+        }
+    }
```
Comment (aleksuss) [inline src/db.rs:1176]:
```suggestion
    }

```

File: src/db_options.rs [hunk 8]
```diff
@@ -133,20 +133,26 @@ pub struct ReadOptions {
     iterate_upper_bound: Option<Vec<u8>>,
 }
 
+pub struct IngestExternalFileOptions {
```
Comment (aleksuss) [inline src/db_options.rs:136]: Almost done. Add a description here please.

File: src/db_options.rs [hunk 9]
```diff
@@ -1861,6 +1873,67 @@ impl Default for ReadOptions {
         }
     }
 }
+impl IngestExternalFileOptions {
```
Comment (DarkEld3r) [inline src/db_options.rs:1876]:
```suggestion

impl IngestExternalFileOptions {
```

**Comment by methyl** [comment]
@aleksuss @DarkEld3r any plans for merging this?

**Comment by aleksuss** [comment]
@methyl Yes. I plan to merge this PR. Add a description [here](https://github.com/rust-rocksdb/rust-rocksdb/pull/421#discussion_r424409519) please.

File: src/db_options.rs (hunk 8 above)
Comment (methyl) [reply src/db_options.rs:137]: Will do, missed that one

**Comment by methyl** [comment]
@aleksuss all clear now I guess

**Pull request merged**